harness/determined/estimator/_reducer.py

```diff
@@ -117,38 +117,102 @@ def cross_slot_reduce(self, per_slot_metrics: List[List[Any]]) -> Any:
+
+
+def _reconstruct_metric(args: List, reconstruct_info: Tuple[type, List]):
+    """Reconstruct lists or dictionaries after passing them through the graph."""
+    metric_type, update_keys = reconstruct_info
+    if metric_type == tf.Tensor:
+        return args[0]
+    if metric_type == list:
+        return args
+    if metric_type == dict:
+        return {k: v for k, v in zip(update_keys, args)}
+    raise AssertionError(f"invalid metric_type: {metric_type}")
+
+
+class _DistributedMetricMaker:
     """
-    _distributed_metric returns a tf.metrics-style tuple of (value_op, update_op).  The value_op is
-    apparently read once after all evaluation is completed, which is where we do the allgather and
-    call the user's cross_slot_reduce to calculate the distributed metric.
+    _DistributedMetricMaker.make_metric() returns a tf.metrics-style tuple of (value_op, update_op).
+    The value_op is read once after all evaluation is completed, which is where we do the allgather
+    and call the user's cross_slot_reduce to calculate the distributed metric.
     """
-    if isinstance(numpy_dtype, tf.dtypes.DType):
-        raise TypeError(f"numpy_dtype parameter must not be a TensorFlow dtype: {numpy_dtype}")
-    np_dtype = np.dtype(numpy_dtype)
-    tf_dtype = tf.compat.v1.as_dtype(numpy_dtype)
-
-    last_accumulate = None  # type: Any
-
-    def py_update(metric: Any) -> None:
-        nonlocal last_accumulate
-        last_accumulate = reducer.accumulate(metric)
-
-    update_op = tf.compat.v1.py_func(py_update, [metric], [])
-
-    def py_value() -> Any:
-        allgathered = context.allgather_metrics(last_accumulate)
-        value = reducer.cross_slot_reduce(allgathered)
-        return np.array(value).astype(np_dtype)
-
-    def build_value_op() -> tf.Operation:
-        return tf.compat.v1.py_func(py_value, [], tf_dtype)
-
-    value_op = context._build_allgather_op(build_value_op)
 
-    return value_op, update_op
+    def __init__(
+        self,
+        context: estimator.EstimatorExperimentalContext,
+        metric: Any,
+        reducer: MetricReducer,
+        numpy_dtype: Any,
+    ) -> None:
+        self.context = context
+        self.reducer = reducer
+
+        # Determined how we are going to pass the metric parameter through the graph, so we can
+        # reconstruct it for the user inside of a py_func.
+        self.update_args, self.reconstruct_info = _deconstruct_metric(metric)
+        self.np_dtype = np.dtype(numpy_dtype)
+        self.tf_dtype = tf.compat.v1.as_dtype(numpy_dtype)
+
+        self.last_accumulate = None
+
+    def _update(self, *args: List) -> None:
+        # Reconstruct the format of the input that the user gave us.
+        metric = _reconstruct_metric(args, self.reconstruct_info)
+        self.last_accumulate = self.reducer.accumulate(metric)
+
+    def _update_op(self) -> tf.Operation:
+        return tf.compat.v1.py_func(self._update, self.update_args, [])
```

Comment:
question: so is the deconstruction / reconstruction necessary because  `tf.compat.v1.py_func` will puke when you pass it a Dict (or anything besides a list of tensors)? If this is the case I know you already called out that you need this for the graph, but calling out `tf.compat.v1.py_func` specifically would be helpful as it took me a non-trivial amount of time to connect the dots here.

Reply:
Yeah, it's that you can pass multiple tensors to a py_func, but if you try to pass a nested structure like a list or dictionary to a py_func, it will try to convert that thing to a single tensor, and that will fail.  So you have to separate the tensors and reconstruct the original thing after the fact.  I'll add this to the docstrings.